lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -1046,12 +1088,36 @@ FIRRTLType SubfieldOp::getResultType(Type inType, StringAttr fieldName,
 
   if (auto flipType = inType.dyn_cast<FlipType>())
     if (auto subType = getResultType(flipType.getElementType(), fieldName, loc))
-      return FlipType::get(subType);
+      return subType;
 
   mlir::emitError(loc, "subfield requires bundle operand");
   return {};
 }
 
+bool SubfieldOp::isFieldFlipped() {
```

Comment:
Curiosity: why do this instead of having the result of the subfield flipped?  I don't understand why you're stripping it above.

Reply:
This moves the type system closer to the SFC and makes type checks and lower types connect/reverse connect selection straightforward.

FIRRTL made this weird decision to make flippedness a property of a bundle field. An example would be the following:
```
wire a: {flip a: UInt<1>}
```

A subfield like `a.a` then has, from the perspective of the SFC, a type of `uint<1>`. This matters for both type checking (e.g., a connect op needs LHS to _exactly_ match RHS by this definition) and for type lowering (e.g., the connect is a reverse connect if the type of the LHS leaf has an odd number of flips). This avoids code that does stuff like:
1) Check if type is flipped
2) If the type is flipped, check if the defining operation was a subfield
3) If so, strip the flip

It's an implementation choice and I struggled with whether to strip flips or keep the current "flip pushing" behavior and then reconstruct SFC semantics when needed. Input at office hour discussions were in favor of just aligning with the SFC behavior because we couldn't find a benefit in diverging and pushing flips was insufficient to replace flow checks.

The only place we then diverge from the SFC is continuing to represent input/output directionality. One cool thing about stripping flips here is that this should disallow an outer flip from ever occurring, except if something is an output port (or input ports if we change the convention).

Reply:
The result of this (which actually addresses your question) is that when walking up definitions to compute flow, you need to capture the field flip. This seemed reasonable given the benefits outlined above.

Reply:
This is a super weird modeling of this.  How does SFC even work if this is the case?  It sounds like no projections of a bundle would ever have flip type, which means that root scalar types work differently then bundles of them together?

Reply:
I think this means that we're going to have to add "output" as a bit on ports if this is true.  This is super wacky.

Reply:
> How does SFC even work if this is the case?

Each field has a parameter that stores flippedness. Something like `flip a: UInt<1>` is stored as:

```scala
BundleType(ArrayBuffer(Field(a,Flip,UIntType(IntWidth(1)))))
```

The type of the subfield is then `UInt<1>`.

> It sounds like no projections of a bundle would ever have flip type, which means that root scalar types work differently then bundles of them together?

Yeah. The type of a subfield is never flipped in the SFC. (I'm assuming projection can't be identity, here?) However, the result may be sink, source, or duplex flow (a function of the declaration kind and the number of flips between the subfield and the declaration, including the field). When the result of a subfield is used in a connect, the flow determines legality of the connection and the type determines connect/reverse connect.

This PR does not have LowerTypes working, yet, but the legality of all the connects in https://github.com/llvm/circt/blob/main/test/Dialect/FIRRTL/lower-types-errors.fir can now be checked.

I admit this is wacky.

> I think this means that we're going to have to add "output" as a bit on ports if this is true. This is super wacky.

I'm still not convinced due to the restrictions on where flip can appear after this PR. A flip can now only show up as the outer flip of an output or on the type of a bundle field. Flips cannot escape bundles to become outer types due to the projection rule above. Therefore, an outer flip type must indicate output.

It _may_ be more efficient to not use `flip` here and just stick a bit in a type, though. Similarly, all the flow and kind information can, I think, be computed in a feedforward manner during construction and make all this fast.

Reply:
Oh I see, yes I agree that flip being a bit in bundle type fields is consistent with SFC.  We can continue to iterate on this in subsequent patches, but this argues that FlipType should go away.